### PR TITLE
Replace webmachine with cowboy

### PR DESCRIPTION
**Pull request by stroop23**

### CODE REVIEW

**Comment by thijsterlouw** [comment]
First thing I do when so many files / lines changed:  search for "test" and I don't see any changed/new tests, which sounds suspicious :)

The change itself looks pretty good. Even though many lines changed, it is pretty isolated and easy to understand.

File: src/piqi_rpc.erl [hunk 1]
```diff
@@ -62,60 +60,52 @@ ensure_started(App) ->
...
     RpcService = normalize_service_def(RpcServiceDef),
     % atomic service addition to both piqi_rpc_monitor and piqi_rpc_http
     try
         % adding Piqi-RPC service first, then adding HTTP resource binding
-        ok = piqi_rpc_monitor:add_service(RpcService),
-        ok = piqi_rpc_http:add_service(RpcService),
-        ServiceDefs = get_service_defs(),
-        ok = set_service_defs([RpcService | ServiceDefs])
+        ok = check_modules(RpcService),
+        ok = add_service_def(RpcService),
```
Comment (thijsterlouw) [inline src/piqi_rpc.erl:77]: add_service_def but not remove_service_def (instead remove_service) , small inconsistency. remove_service internally calls remove_service_def.

File: src/piqi_rpc_http.erl [hunk 2]
```diff
@@ -21,77 +21,74 @@
...
+        {'_', lists:flatten(Routes)}
+    ],
+    cowboy_router:compile(Dispatch).
+
+rpc_service_to_cowboy_route(_RpcService = {ImplMod, RpcMod, UrlPath, Options}) ->
+    LAdjustedPath = case string:left(UrlPath, 1) of
+        "/" -> UrlPath;
+        _ -> "/" ++ UrlPath
+    end,
+    AdjustedPath = case string:right(LAdjustedPath, 1) of
```
Comment (thijsterlouw) [inline src/piqi_rpc_http.erl:56]: is this behaviour (of fixing the path by pre/post-pending slashes) consistent with the old behaviour ?

File: src/piqi_rpc_resource.erl [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+            reply(411, [], <<"non-empty input expected">>, Req, State)
+    end.
+
+rpc(Req, State, InputFormat, OutputFormat, <<>>) when InputFormat =/= 'pb' ->
+    % Body must be undefined if empty, unless it's protobuf
+    rpc(Req, State, InputFormat, OutputFormat, undefined);
+rpc(Req, State = #rpc_state{rpc_mod = RpcMod, impl_mod =ImplMod, options = Options}, InputFormat, OutputFormat, Body) ->
+    {Function, Req2} = cowboy_req:binding(function, Req),
+    % Pass the whole request through the process dictionary
+    % @TODO based on options pass it or parts as second argument to the implementation function call.
```
Comment (thijsterlouw) [inline src/piqi_rpc_resource.erl:84]: is this TODO important ?

**Comment by thijsterlouw** [comment]
Besides that, LGTM, nice one!

File: src/piqi_rpc.erl (hunk 1 above)
Comment (stroop23) [inline src/piqi_rpc.erl:77]: ? 

add_service does add_service_def, same as remove_service does remove_service_def ?

File: src/piqi_rpc_http.erl (hunk 2 above)
Comment (stroop23) [inline src/piqi_rpc_http.erl:56]: i believe so, i think the tests failed without this. But i'll double check.

File: src/piqi_rpc_resource.erl (hunk 3 above)
Comment (stroop23) [inline src/piqi_rpc_resource.erl:84]: Yes, however the phrasing is wrong. It should state that we need to find a better way to pass request related parameters to the implementation modules.

**Comment by yfyf** [comment]
It's hard to compare the `piqi_rpc_resource` due to the remove / add. Maybe you could rebase so that there's an actual diff?

Otherwise seems ok, but definitely needs at least some basic tests.

**Pull request merged**